types/statuses/index.d.ts

```diff
@@ -3,13 +3,29 @@
 // Definitions by: Tanguy Krotoff <https://github.com/tkrotoff>
 //                 BendingBender <https://github.com/BendingBender>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 4.7
+
+type NumericAscii = '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '0';
+type NonNumericAscii<S> = S extends `${NumericAscii}` ? never : any;
+
+type IsNumericString<S extends string> = S extends `${number}` ? any : never;
+
+type IsNonNumericString<S extends string> = S extends `${NonNumericAscii<S>}${infer _}` ? any : never;
+
+type Result<S extends string> =
+    S extends IsNumericString<S>
+    ? string
+    : S extends IsNonNumericString<S>
+    ? number
+    : string|number;
 
 export = status;
 
-declare const status: Status;
+declare const status: status;
 
-interface Status {
-    (code_msg: number | string): number | string;
+interface status {
+    (code: number): string;
+    <S extends string>(code: S): Result<S>;
```

Comment:
Now you have the type `Result` defined twice. Just delete the one outside of the `namespace status` and use the one from the namespace here:
```suggestion
    <S extends string>(code: S): status.Result<S>;
```